File: saas/lib/billing/rate-card.ts
import type { ModelRate } from "./net-of-cache";

// Fallback rate card mirroring supabase/seed.sql. Authoritative billing reads
// model_pricing (+ per-tenant overrides) from the DB; this is the default when
// a tenant has no override row. $/token = ($/1M) / 1e6.
const M = 1e6;
const r = (base: number, cached?: number): ModelRate => ({
  baseInputRate: base / M,
  cachedInputRate: (cached ?? base) / M,
});

export const RATE_CARD: Record<string, ModelRate> = {
  "gpt-5.5": r(5.0, 0.5),
  "gpt-5.4": r(2.5, 0.25),
  "gpt-5.4-mini": r(0.75, 0.075),
  "gpt-5.2-codex": r(1.75),
  "gpt-5": r(1.25),
  "gpt-5-mini": r(0.25),
  "gpt-4o": r(2.5),
  "claude-opus-4-8": r(5.0, 0.5),
  "claude-opus-4-7": r(5.0, 0.5),
  "claude-sonnet-4-6": r(3.0, 0.3),
  "claude-haiku-4-5": r(1.0, 0.1),
  "claude-fable-5": r(10.0, 1.0),
  "gemini-3.1-pro": r(2.0),
  "gemini-3.5-flash": r(1.5, 0.15),
  "gemini-2.5-pro": r(1.25),
  "grok-4": r(3.0),
  "grok-4-fast": r(0.2),
};

File: agent/wisp_agent/rate_card.py
"""Local model rate card (mirror of the SaaS `model_pricing` seed).

Used only for the on-device glanceable dollar figure. The authoritative billing
rate lives server-side; this local copy must stay in sync with
supabase/migrations seed. Rates are $/token (per-million figures / 1e6).

Where a provider's cache rate is unknown we default cached == base
(conservative: assumes no cache discount).
"""

from __future__ import annotations

from .billing import ModelRate

_PER_M = 1_000_000


def _rate(base_per_m: float, cached_per_m: float | None = None) -> ModelRate:
    return ModelRate(
        base_input_rate=base_per_m / _PER_M,
        cached_input_rate=(cached_per_m if cached_per_m is not None else base_per_m) / _PER_M,
    )


# $/1M tokens (base, cached). June 2026 list prices.
DEFAULT_RATE_CARD: dict[str, ModelRate] = {
    # OpenAI
    "gpt-5.5": _rate(5.00, 0.50),
    "gpt-5.4": _rate(2.50, 0.25),
    "gpt-5.4-mini": _rate(0.75, 0.075),
    "gpt-5.2-codex": _rate(1.75),
    "gpt-5": _rate(1.25),
    "gpt-5-mini": _rate(0.25),
    "gpt-4o": _rate(2.50),
    # Anthropic
    "claude-opus-4-8": _rate(5.00, 0.50),
    "claude-opus-4-7": _rate(5.00, 0.50),
    "claude-sonnet-4-6": _rate(3.00, 0.30),
    "claude-haiku-4-5": _rate(1.00, 0.10),
    "claude-fable-5": _rate(10.00, 1.00),
    # Google
    "gemini-3.1-pro": _rate(2.00),
    "gemini-3.5-flash": _rate(1.50, 0.15),
    "gemini-2.5-pro": _rate(1.25),
    # xAI
    "grok-4": _rate(3.00),
    "grok-4-fast": _rate(0.20),
}


def resolve(model: str) -> ModelRate | None:
    return DEFAULT_RATE_CARD.get(model)
